[React] Add PWA features

samuelmeuli opened the pull request:
This PR adds PWA features to the React app. All PWA-related Lighthouse checks will pass with these changes.

**PWA-related changes:**

- Improved the web app manifest with theme colors and icon links
- Added favicons in different sizes
- Added a custom service-worker setup. The default one generated by `create-react-app` is not sufficient because it only covers files which are part of the React app, i.e. not AEM resources that are requested from other paths

I've tried to add as many code comments as possible, so the implementation details should hopefully be self-explanatory. @godanny86, please let me know if I can clarify anything :)

**Other required changes:**

- Switched from `node-sass` to `sass` (Dart implementation) for compatibility with the latest macOS version
- Updated `react-scripts` (`create-react-app`) and `aem-clientlib-generator` because the newer versions include changes that allow code splitting and better SPA asset handling with AEM ClientLibs
- Updated the `aem-clientlib-generator` config to the latest one from the archetype (same reason)

samuelmeuli committed:
Add PWA features

samuelmeuli committed:
Rename service worker registration file

samuelmeuli committed:
Improve code reuse

samuelmeuli committed:
Update favicon

samuelmeuli committed:
Rename favicons

samuelmeuli committed:
[React] Use external module for service worker registration

samuelmeuli committed:
[React] Fix HTML header

godanny86 commented on react-app/package.json:
```diff
@@ -10,24 +10,26 @@
     "styleguide": "styleguidist server",
     "styleguide:build": "styleguidist build"
   },
+  "homepage": "/etc.clientlibs/wknd-events/clientlibs/react-app/resources",
```
@samuelmeuli -> this seems a bit strange..is it just to facilitate relative path resolution?

samuelmeuli replied:
This path is prepended to the files in `asset-manifest.json` and `precache-manifest.json`, i.e. it's required for code splitting and caching to work. But maybe it's less confusing if the `PUBLIC_URL` env variable is used instead of the `homepage` key

samuelmeuli replied:
I've changed this and documented why it's required

godanny86 replied:
nice, i like it. thanks @samuelmeuli!

samuelmeuli committed:
Use `PUBLIC_URL` env variable instead of "homepage" key

godanny86 reviewed (approved):
@samuelmeuli I noticed that the site works completely offline, very cool! One other (hopefully small) feature would be to alert the user that they are browsing offline with a small banner?

samuelmeuli committed:
Add offline banner component

samuelmeuli commented:
Yes, that's a good idea. I've added the banner component to both PRs :)

samuelmeuli committed:
Delete precache manifest file before generating service worker

The precache manifest file is generated by Workbox v4 (used in `react-scripts`). Since v5 (used for generating the custom service worker), this is no longer output as a separate file. It therefore makes sense to remove this file after the `create-react-app` build in order to avoid confusion

samuelmeuli committed:
Remove ServiceWorkerHeaderFilter

samuelmeuli committed:
Add `dispatcher` module

samuelmeuli committed:
Add service worker header using Dispatcher

samuelmeuli commented:
@godanny86 As discussed, I've replaced the `ServiceWorkerHeaderFilter` Sling filter with a custom Dispatcher config (copied from WKND Sites and extended by a rule which adds the `Service-Worker-Allowed` header to the `service-worker.js` file, see a42cadd).
